alibi_detect/cd/metrics.py

```diff
@@ -0,0 +1,333 @@
+        the detector.
+    C
+        A reservoir of context data. On each run instances are randomly subsampled to provide reference and test sets
+        for the detector. Should only be given for the :class:`~alibi_detect.cd.ContextMMDDrift` detector.
+    model
+        A PyTorch, TensorFlow or Sklearn model. Should only be given for detectors accepting a `model` arg, such
+        as :class:`~alibi_detect.cd.ClassifierDrift`.
+    detector_kwargs
+        Keyword arguments to pass to detector.
+    n_runs
+        The number of experiment runs. A larger number will give more stable p-value distributions, at the expense of
+        a longer runtime.
+    n_samples
+        The number of samples to randomly sample from the data reservoir to create reference and test data for each run.
+        The two data subsampled data sets will each contain `n_samples` instances, such that `2*n_samples` instances
+        are sampled in total.
+    qq_plot
+        Whether to visualise the p-values on a Quantile-Quantile plot. This can also be done later with the
+        :func:`~plot_qq` function.
+    hist_plot
+        Whether to visualise the p-values on a histogram. This can also be done later with the :func:`~plot_hist`
+        function.
+    random_seed
+        Random seed (or :class:`~np.random.Generator`) to use for random sampling. If `None`, then fresh,
+        unpredictable entropy will be pulled from the OS.
+
+    Returns
+    -------
+    A numpy array containing the detector's p-value prediction for each of the `n_runs` number of runs.
+    """
+    # Check data sizes
+    n_data = len(X)
+    _check_sufficient_data_size(X, n_samples, n_runs)
+
+    # Setup detector_kwargs
+    if detector_kwargs is None:
+        detector_kwargs = {}
+
+    # Perform preprocessing (if preprocess_fn exists) to save repeated compute in loop
+    preprocess_fn = detector_kwargs.pop('preprocess_fn', None)
+    if preprocess_fn is not None:
+        X = preprocess_fn(X)
+
+    # NumPy RNG
+    rng = np.random.default_rng(random_seed)
+
+    # Main experiment loop
+    p_vals_list = []
+    for _ in tqdm(range(n_runs)):
+        # Subsample data
+        idx = rng.choice(n_data, size=n_data, replace=False)
+        idx_ref, idx_nodrift = idx[:n_samples], idx[n_samples:2*n_samples]
+        x_ref, x_nodrift = X[idx_ref], X[idx_nodrift]
+        detector_args, predict_args = [x_ref], [x_nodrift]
+
+        if C is not None:
+            c_ref, c_nodrift = C[idx_ref], C[idx_nodrift]
+            detector_args.append(c_ref)
+            predict_args.append(c_nodrift)
+
+        if model is not None:
+            detector_args.append(model)
+
+        # Init detector and predict
+        dd = detector(*detector_args, **detector_kwargs)
+        preds = dd.predict(*predict_args)
+        p_vals_list.append(preds['data']['p_val'])
+    p_vals = np.array(p_vals_list).flatten()
+    # TODO - is above OK? (to flatten over features for univariate detectors)
```

Comment:
If so I think we should specify the calibration is at the feature rather than detector level. Alternatively we could have an arg corresponding to the correction and show compute the calibration at the detector level.

Reply:
I was thinking more along the lines of return the p-values as given when there is a p-value per feature and optionally allowing them to aggregate those multiple p-values using a correction to obtain instance level p-values.

Reply:
Thanks @ojcobb. So have `correction=None` as default and return the p-values as given (i.e. per feature) in that instance. Otherwise, aggregate to give instance-level p-values if `correction` is given?